fenicsprecice/adapter_core.py

```diff
@@ -274,3 +354,186 @@ def get_forces_as_point_sources(fixed_boundary, function_space, coupling_mesh_ve
+        send_reqs.append(req)
+
+    # Wait for all non-blocking communication operations to complete
+    MPI.Request.Waitall(send_reqs)
+
+    nonownership_data = dict()
+    # Collect received ownership data of other ranks
+    counter = 0
+    for neigh in neigh_ranks:
+        nonownership_data[neigh] = recv_reqs[counter].wait()
+        counter += 1
+
+    # Create dictionary: keys are Global IDs of vertices not owned by this rank, values are ranks from which
+    # data on these vertices needs to be received
+    send_data = dict()
+    for neigh in neigh_ranks:
+        for uo_gid in nonownership_data[neigh]:
+            for o_gid in owned_gids:
+                if uo_gid == o_gid:
+                    send_data[o_gid] = neigh
+
+    return send_data, recv_data
+
+
+def communicate_shared_vertices(comm, rank, fenics_gids, owned_coords, fenics_coords, owned_data,
+                                send_pts, recv_pts):
+    """
+
+    Parameters
+    ----------
+    comm : Object of class MPI.COMM_WORLD from mpi4py package
+        A predefined intracommunicator instance available in mpi4py.
+    rank : int
+        Rank of calling process in a communicator obtained from MPI.
+    fenics_gids : numpy array
+        Array of global indices of vertices on the coupling interface as seen by this rank.
+    owned_coords : array_like
+            The coordinates of the vertices on the coupling interface and owned by this rank, in a numpy array [N x D]
+            where N = number of vertices and D = dimensions of geometry.
+    fenics_coords : array_like
+        The coordinates of the vertices on the coupling interface as seen by this rank, in a numpy array [N x D] where
+        N = number of vertices and D = dimensions of geometry.
+    owned_data : dict_like
+        The coupling data. A dictionary containing nodal data with vertex coordinates as key and associated data as
+            values.
+    send_pts : dict_like
+        A dictionary with global IDs of vertices whose data needs to be sent as the keys, and the ranks to which the
+        data needs to be sent as the values.
+    recv_pts : dict_like
+        A dictionary with global IDs of vertices on which data needs to be received as keys, and the ranks from which
+        the data needs to be received as values.
+
+    Returns
+    -------
+    fenics_data : array_like
+        The updated coupling data. A dictionary containing nodal data with vertex coordinates of all seen vertices as
+        the keys, and the associated data as values.
+
+    """
+    # Create fenics type shared data array for communication
+    fenics_data = dict.fromkeys(tuple(key) for key in fenics_coords)
+
+    # Attach data read from preCICE to appropriate ids in FEniCS style array (which includes duplicates)
+    for coord in fenics_coords:
+        for owned_coord in owned_coords:
+            if (coord == owned_coord).all():
+                fenics_data[tuple(coord)] = owned_data[tuple(owned_coord)]
```

Comment:
See comment above: We can just skip this, if we directly work with `owned_data` (maybe let's call it `data` then). Or am I missing something?

Reply:
After short discussion: Let's create a copy of `owned_data` called `fenics_data` and then append the missing keys `fenics_data[key] = None`. Due to security reasons, we don't want to modify `owned_data` in-place.